### ISSUE TITLE
Panic seen in isAlertRuleMatching after upgrading to v2.5.2

### ISSUE THREAD

**Issue by superseb** [rancher/rancher#29518]
From https://github.com/rancher/rancher/issues/29460#issuecomment-708278892:

I believe I have the same issue with Rancher 2.5.2 (standalone Docker container), used to manage a single bare-metal cluster (K8s v1.19.2-rancher1-1).

@rbq If you can provide more info on this, that would help. We can fix the panic based on code but any info would be nice.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x289aa4a]

goroutine 7030 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/apimachinery@v0.19.0-rancher1/pkg/util/runtime/runtime.go:55 +0x105
panic(0x3c138c0, 0x73706a0)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/rancher/rancher/pkg/controllers/managementuser/alert/watcher.(*ClusterScanWatcher).isAlertRuleMatching(...)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementuser/alert/watcher/cluster_scan.go:110
github.com/rancher/rancher/pkg/controllers/managementuser/alert/watcher.(*ClusterScanWatcher).Sync(0xc00d96bc70, 0xc01224f830, 0x22, 0xc0222d2300, 0x44c3ef8, 0x4, 0x44d3772, 0xb)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementuser/alert/watcher/cluster_scan.go:79 +0x30a
github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3.(*clusterScanController).AddClusterScopedHandler.func1(0xc01224f830, 0x22, 0x442dfa0, 0xc0222d2300, 0xc02224fb20, 0x3, 0x3, 0x0)
	/go/src/github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3/zz_generated_cluster_scan_controller.go:182 +0x79
github.com/rancher/norman/controller.(*genericController).AddHandler.func1(0xc01224f830, 0x22, 0x4e247a0, 0xc0222d2300, 0x4e247a0, 0xc0222d2300, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20200930000340-693d65aaffe3/controller/generic_controller.go:62 +0x221
github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0xc00eb509f0, 0xc01224f830, 0x22, 0x4e247a0, 0xc0222d2300, 0x4e247a0, 0xc0222d2300, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/sharedcontroller.go:29 +0x4e
github.com/rancher/lasso/pkg/controller.(*sharedHandler).OnChange(0xc00c375ad0, 0xc01224f830, 0x22, 0x4e247a0, 0xc0222d2300, 0xc00e960301, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/sharedhandler.go:65 +0x14a
github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc000bfb290, 0xc01224f830, 0x22, 0xc00e95e840, 0xc015946d50)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/controller.go:210 +0xd1
github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc000bfb290, 0x390a700, 0xc02167aec0, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/controller.go:192 +0xd7
github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc000bfb290, 0x203003)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/controller.go:169 +0x51
github.com/rancher/lasso/pkg/controller.(*controller).runWorker(0xc000bfb290)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/controller.go:158 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00e958c10)
	/go/pkg/mod/github.com/rancher/apimachinery@v0.19.0-rancher1/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00e958c10, 0x4dfa100, 0xc00d07ee10, 0x3a22636570732201, 0xc00153de00)
	/go/pkg/mod/github.com/rancher/apimachinery@v0.19.0-rancher1/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00e958c10, 0x3b9aca00, 0x0, 0x3136386664623101, 0xc00153de00)
	/go/pkg/mod/github.com/rancher/apimachinery@v0.19.0-rancher1/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc00e958c10, 0x3b9aca00, 0xc00153de00)
	/go/pkg/mod/github.com/rancher/apimachinery@v0.19.0-rancher1/pkg/util/wait/wait.go:90 +0x4d
created by github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20200905045615-7fcb07d6a20b/pkg/controller/controller.go:129 +0x31f
```

**Comment by rbq** [comment]
Ooops, sorry for mixing this up with the unrelated issue. I'll try to provide some additional information:

> the upgrade path and how many/what clusters you are managing and if you are using cluster scans/alerting

Honestly, it's a mess and I clearly tried to change too many things at once. I was running a single container installation with Rancher 2.4.2 managing a single cluster (K8s 1.18.2 I believe). After replacing the Rancher instance with a 2.5.0 container I upgraded the cluster to 1.19.2 and tried to migrate Istio (worked), alerting (didn't work), and monitoring (kind-of worked) from the old management UI to new charts via Cluster Explorer. That involved a bit of manually deleting resources (incl. namespaces) as the old monitoring didn't quite want to go away completely.

The system seemed to run at first, but periodically crashed. At first I assumed it was a memory issue and unsuccessfully tried to grow the VM from 4 to 8 GB, then 12 GB. I think next 2.5.1 came out, I tried to upgrade the container but it wouldn't even start up (etcd reported waiting for connections on the expected ports but Rancher couldn't connect at all). After not being able to debug or resolve the problem I assumed some kind of volume/etcd corruption after all the crashing and forced restarting. I therefore removed the Rancher container and created a new one (Rancher 2.5.1), restoring my backup of the Rancher 2.4.8 data. Then I changed the K8s version (which was being shown as 1.18 due to the restored Rancher backup) back to 1.19 and tried migrating Istio and monitoring again.

**Comment by nickgerace** [comment]
Want to make sure that we do not stray off the path here, but from following the stack trace there may be a few good places to look...
- [wait.go line 140](https://github.com/kubernetes/apimachinery/blob/v0.19.2/pkg/util/wait/wait.go#L140)
- [controller.go line 157-177](https://github.com/rancher/lasso/blob/7fcb07d6a20b6fc288b1fc9acff8e4be8706773f/pkg/controller/controller.go#L157-L177)

**Comment by nickgerace** [comment]
Created a PR for this. Essentially, the ```CisScanStatus``` was not created because the ```CisScan``` either did not complete, or it failed. The panic was a result of the ```isAlertRuleMatching``` function trying to access fields for the non-existent ```CisScanStatus``` object.

**Comment by sowmyav27** [comment]
Reopening for a fix in 2.5 branch

**Comment by izaac** [comment]
Rancher version: `v2.5-head (10/22/2020)` _1599dc1_
Rancher version: `v2.4-head (10/22/2020)` _6ab3cb450_

Executed P1 cases using alerts/notifiers (monitoring is also a dependency for this).
I was able to run manual scans and recieve alerts for failed and success scans.
I received notifications/alerts from executed scheduled scans, both from failed and success scans.

Slack was used as the notifier.
